File: v1/form_renderer.go
package smartform

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/juicycleff/smartform/v1/template"
)

// FormRenderer converts form schemas to JSON representations for the frontend
type FormRenderer struct {
	schema         *FormSchema
	templateEngine *template.TemplateEngine
}

// NewFormRenderer creates a new form renderer
func NewFormRenderer(schema *FormSchema) *FormRenderer {
	return &FormRenderer{
		schema:         schema,
		templateEngine: template.NewTemplateEngine(),
	}
}

// RenderJSON converts the form schema to a JSON string
func (fr *FormRenderer) RenderJSON() (string, error) {
	data, err := json.MarshalIndent(fr.schema, "", "  ")
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// RenderJSONWithContext renders the form with context-specific modifications
func (fr *FormRenderer) RenderJSONWithContext(context map[string]interface{}) (string, error) {
	// Create a copy of the schema to modify
	schemaCopy := fr.copySchemaWithContext(context)

	// Convert to JSON
	data, err := json.MarshalIndent(schemaCopy, "", "  ")
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// copyFieldWithContext creates a context-aware copy of a field
func (fr *FormRenderer) copyFieldWithContext(field *Field, context map[string]interface{}) *Field {
	// Create a new field with the same basic properties
	fieldCopy := &Field{
		ID:              field.ID,
		Type:            field.Type,
		Label:           field.Label,
		Required:        field.Required,
		DefaultValue:    field.DefaultValue,
		Placeholder:     field.Placeholder,
		HelpText:        field.HelpText,
		Order:           field.Order,
		ValidationRules: make([]*ValidationRule, len(field.ValidationRules)),
		Properties:      make(map[string]interface{}),
		Nested:          []*Field{},
	}

	fieldCopy.Label = fr.evaluateTemplateString(field.Label, context)
	fieldCopy.Placeholder = fr.evaluateTemplateString(field.Placeholder, context)
	fieldCopy.HelpText = fr.evaluateTemplateString(field.HelpText, context)

	// Handle DefaultWhen conditions
	if field.DefaultWhen != nil && len(field.DefaultWhen) > 0 {
		validator := NewValidator(fr.schema)
		for _, defaultWhen := range field.DefaultWhen {
			if validator.evaluateCondition(defaultWhen.Condition, context) {
				// Evaluate the default value if it's a template expression
				if strValue, ok := defaultWhen.Value.(string); ok && fr.containsTemplateExpression(strValue) {
					evaluatedValue, err := fr.templateEngine.EvaluateExpression(strValue, context)
					if err == nil {
						fieldCopy.DefaultValue = evaluatedValue
					} else {
						fieldCopy.DefaultValue = defaultWhen.Value
					}
				} else {
					fieldCopy.DefaultValue = defaultWhen.Value
				}
				break
			}
		}
	} else if field.DefaultValue != nil {
		// Evaluate the default value if it's a template expression
		if strValue, ok := field.DefaultValue.(string); ok && fr.containsTemplateExpression(strValue) {
			evaluatedValue, err := fr.templateEngine.EvaluateExpression(strValue, context)
			if err == nil {
				fieldCopy.DefaultValue = evaluatedValue
			}
		}
	}

	// Handle requiredIf condition
	if field.RequiredIf != nil {
		fieldCopy.RequiredIf = fr.copyCondition(field.RequiredIf)
	}

	// Copy validation rules
	for i, rule := range field.ValidationRules {
		fieldCopy.ValidationRules[i] = &ValidationRule{
			Type:       rule.Type,
			Message:    rule.Message,
			Parameters: rule.Parameters,
		}
	}

	// Copy properties
	for k, v := range field.Properties {
		fieldCopy.Properties[k] = v
	}

	// Handle visibility condition
	if field.Visible != nil {
		fieldCopy.Visible = fr.copyCondition(field.Visible)
	}

	// Handle enablement condition
	if field.Enabled != nil {
		fieldCopy.Enabled = fr.copyCondition(field.Enabled)

		// Evaluate if field should be disabled in this context
		validator := NewValidator(fr.schema)
		if !validator.evaluateCondition(field.Enabled, context) {
			fieldCopy.Properties["disabled"] = true
		}
	}

	// Handle options for select-type fields
	if field.Options != nil {
		fieldCopy.Options = fr.copyOptionsWithContext(field.Options, context)
	}

	// Handle nested fields
	if field.Nested != nil {
		for _, nestedField := range field.Nested {
			// Skip nested fields that aren't visible in this context
			if nestedField.Visible != nil {
				validator := NewValidator(fr.schema)
				if !validator.evaluateCondition(nestedField.Visible, context) {
					continue
				}
			}

			nestedCopy := fr.copyFieldWithContext(nestedField, context)
			fieldCopy.Nested = append(fieldCopy.Nested, nestedCopy)
		}
	}

	return fieldCopy
}

// copySchemaWithContext creates a context-aware copy of the schema
func (fr *FormRenderer) copySchemaWithContext(context map[string]interface{}) *FormSchema {
	// Create a new schema with the same basic properties
	schemaCopy := &FormSchema{
		ID:          fr.schema.ID,
		Title:       fr.schema.Title,
		Description: fr.schema.Description,
		Fields:      []*Field{},
		Properties:  make(map[string]interface{}),
	}

	// Copy over properties
	for k, v := range fr.schema.Properties {
		schemaCopy.Properties[k] = v
	}

	// Process fields based on context
	for _, field := range fr.schema.Fields {
		// Skip fields that should not be visible in this context
		if field.Visible != nil {
			// Create a validator to evaluate the condition
			validator := NewValidator(fr.schema)
			if !validator.evaluateCondition(field.Visible, context) {
				continue
			}
		}

		// Include the field with possible context-specific modifications
		fieldCopy := fr.copyFieldWithContext(field, context)
		schemaCopy.Fields = append(schemaCopy.Fields, fieldCopy)
	}

	// Sort fields by order
	schemaCopy.SortFields()

	return schemaCopy
}

// copyCondition creates a copy of a condition
func (fr *FormRenderer) copyCondition(condition *Condition) *Condition {
	if condition == nil {
		return nil
	}

	conditionCopy := &Condition{
		Type:       condition.Type,
		Field:      condition.Field,
		Value:      condition.Value,
		Operator:   condition.Operator,
		Expression: condition.Expression,
	}

	// Copy nested conditions
	if condition.Conditions != nil {
		conditionCopy.Conditions = make([]*Condition, len(condition.Conditions))
		for i, subCondition := range condition.Conditions {
			conditionCopy.Conditions[i] = fr.copyCondition(subCondition)
		}
	}

	return conditionCopy
}

// copyOptionsWithContext creates a context-aware copy of field options
func (fr *FormRenderer) copyOptionsWithContext(options *OptionsConfig, context map[string]interface{}) *OptionsConfig {
	if options == nil {
		return nil
	}

	optionsCopy := &OptionsConfig{
		Type: options.Type,
	}

	// Copy static options
	if options.Static != nil {
		optionsCopy.Static = make([]*Option, len(options.Static))
		for i, option := range options.Static {
			optionsCopy.Static[i] = &Option{
				Value: option.Value,
				Label: option.Label,
				Icon:  option.Icon,
			}
		}
	}

	// Handle dynamic options source
	if options.DynamicSource != nil {
		optionsCopy.DynamicSource = &DynamicSource{
			Type:      options.DynamicSource.Type,
			Endpoint:  options.DynamicSource.Endpoint,
			Method:    options.DynamicSource.Method,
			ValuePath: options.DynamicSource.ValuePath,
			LabelPath: options.DynamicSource.LabelPath,
			RefreshOn: make([]string, len(options.DynamicSource.RefreshOn)),
		}

		// Copy refresh triggers
		copy(optionsCopy.DynamicSource.RefreshOn, options.DynamicSource.RefreshOn)

		// Copy headers
		if options.DynamicSource.Headers != nil {
			optionsCopy.DynamicSource.Headers = make(map[string]string)
			for k, v := range options.DynamicSource.Headers {
				optionsCopy.DynamicSource.Headers[k] = v
			}
		}

		// Copy parameters
		if options.DynamicSource.Parameters != nil {
			optionsCopy.DynamicSource.Parameters = make(map[string]interface{})
			for k, v := range options.DynamicSource.Parameters {
				optionsCopy.DynamicSource.Parameters[k] = v
			}
		}
	}

	// Handle dependent options
	if options.Dependency != nil {
		// Process dependent options based on context
		dependentField := options.Dependency.Field
		dependentValue := fr.getValueFromContext(context, dependentField)

		optionsCopy.Dependency = &OptionsDependency{
			Field: dependentField,
		}

		// Copy value map
		if options.Dependency.ValueMap != nil {
			optionsCopy.Dependency.ValueMap = make(map[string][]*Option)
			for k, v := range options.Dependency.ValueMap {
				optsCopy := make([]*Option, len(v))
				for i, opt := range v {
					optsCopy[i] = &Option{
						Value: opt.Value,
						Label: opt.Label,
						Icon:  opt.Icon,
					}
				}
				optionsCopy.Dependency.ValueMap[k] = optsCopy
			}

			// Filter options based on dependent field value
			if dependentValue != nil {
				valueStr := fmt.Sprintf("%v", dependentValue)
				if filteredOptions, ok := options.Dependency.ValueMap[valueStr]; ok {
					optionsCopy.Static = make([]*Option, len(filteredOptions))
					for i, opt := range filteredOptions {
						optionsCopy.Static[i] = &Option{
							Value: opt.Value,
							Label: opt.Label,
							Icon:  opt.Icon,
						}
					}
				}
			}
		}
	}

	return optionsCopy
}

// getValueFromContext gets a value from the context using dot notation
func (fr *FormRenderer) getValueFromContext(context map[string]interface{}, path string) interface{} {
	validator := NewValidator(fr.schema)
	return validator.getValueByPath(context, path)
}

// evaluateTemplateString evaluates a string that may contain template expressions
func (fr *FormRenderer) evaluateTemplateString(input string, context map[string]interface{}) string {
	if input == "" || !fr.containsTemplateExpression(input) {
		return input
	}

	result, err := fr.templateEngine.EvaluateExpressionAsString(input, context)
	if err != nil {
		return input
	}
	return result
}

// containsTemplateExpression checks if a string contains template expressions
func (fr *FormRenderer) containsTemplateExpression(input string) bool {
	return strings.Contains(input, "${")
}

File: react/core/form-renderer.ts
import type { Condition, Field, FormSchema } from "./types";
import { Validator } from "./validation-engine";

/**
 * FormRenderer converts form schemas to JSON representations for the frontend
 */
export class FormRenderer {
  private schema: FormSchema;
  private validator: Validator;

  /**
   * Creates a new form renderer
   * @param schema Form schema to render
   */
  constructor(schema: FormSchema) {
    this.schema = schema;
    this.validator = new Validator(schema);
  }

  /**
   * Converts the form schema to a JSON string
   */
  renderJson(): string {
    return JSON.stringify(this.schema, null, 2);
  }

  /**
   * Renders the form with context-specific modifications
   * @param context Context data
   */
  renderJsonWithContext(context: Record<string, any>): string {
    // Create a copy of the schema to modify
    const schemaCopy = this.copySchemaWithContext(context);

    // Convert to JSON
    return JSON.stringify(schemaCopy, null, 2);
  }

  /**
   * Creates a context-aware copy of a field
   * @param field Field to copy
   * @param context Context data
   */
  private copyFieldWithContext(
    field: Field,
    context: Record<string, any>,
  ): Field {
    // Create a new field with the same basic properties
    const fieldCopy: Field = {
      id: field.id,
      type: field.type,
      label: field.label,
      required: field.required,
      defaultValue: field.defaultValue,
      placeholder: field.placeholder,
      helpText: field.helpText,
      order: field.order,
      properties: { ...field.properties },
      nested: [],
    };

    // Handle requiredIf condition
    if (field.requiredIf) {
      fieldCopy.requiredIf = this.copyCondition(field.requiredIf);
    }

    // Copy validation rules
    if (field.validationRules) {
      fieldCopy.validationRules = field.validationRules.map((rule) => ({
        type: rule.type,
        message: rule.message,
        parameters: rule.parameters,
      }));
    }

    // Handle visibility condition
    if (field.visible) {
      fieldCopy.visible = this.copyCondition(field.visible);
    }

    // Handle enablement condition
    if (field.enabled) {
      fieldCopy.enabled = this.copyCondition(field.enabled);

      // Evaluate if field should be disabled in this context
      if (!this.validator.evaluateCondition(field.enabled, context)) {
        fieldCopy.properties!.disabled = true;
      }
    }

    // Handle options for select-type fields
    if (field.options) {
      fieldCopy.options = this.copyOptionsWithContext(field.options, context);
    }

    // Handle nested fields
    if (field.nested && field.nested.length > 0) {
      for (const nestedField of field.nested) {
        // Skip nested fields that aren't visible in this context
        if (
          nestedField.visible &&
          !this.validator.evaluateCondition(nestedField.visible, context)
        ) {
          continue;
        }

        const nestedCopy = this.copyFieldWithContext(nestedField, context);
        fieldCopy.nested?.push(nestedCopy);
      }
    }

    return fieldCopy;
  }

  /**
   * Creates a context-aware copy of the schema
   * @param context Context data
   */
  private copySchemaWithContext(context: Record<string, any>): FormSchema {
    // Create a new schema with the same basic properties
    const schemaCopy: FormSchema = {
      id: this.schema.id,
      title: this.schema.title,
      description: this.schema.description,
      type: this.schema.type,
      authType: this.schema.authType,
      fields: [],
      properties: { ...this.schema.properties },
    };

    // Process fields based on context
    for (const field of this.schema.fields) {
      // Skip fields that should not be visible in this context
      if (
        field.visible &&
        !this.validator.evaluateCondition(field.visible, context)
      ) {
        continue;
      }

      // Include the field with possible context-specific modifications
      const fieldCopy = this.copyFieldWithContext(field, context);
      schemaCopy.fields.push(fieldCopy);
    }

    // Sort fields by order
    schemaCopy.fields.sort((a, b) => a.order - b.order);

    return schemaCopy;
  }

  /**
   * Creates a copy of a condition
   * @param condition Condition to copy
   */
  private copyCondition(condition: Condition): Condition {
    const conditionCopy: Condition = {
      type: condition.type,
      field: condition.field,
      value: condition.value,
      operator: condition.operator,
      expression: condition.expression,
    };

    // Copy nested conditions
    if (condition.conditions && condition.conditions.length > 0) {
      conditionCopy.conditions = condition.conditions.map((subCondition) =>
        this.copyCondition(subCondition),
      );
    }

    return conditionCopy;
  }

  /**
   * Creates a context-aware copy of field options
   * @param options Options to copy
   * @param context Context data
   */
  private copyOptionsWithContext(
    options: any,
    context: Record<string, any>,
  ): any {
    if (!options) {
      return null;
    }

    const optionsCopy = {
      type: options.type,
    } as any;

    // Copy static options
    if (options.static) {
      optionsCopy.static = options.static.map((option: any) => ({
        value: option.value,
        label: option.label,
        icon: option.icon,
      }));
    }

    // Handle dynamic options source
    if (options.dynamicSource) {
      optionsCopy.dynamicSource = {
        type: options.dynamicSource.type,
        endpoint: options.dynamicSource.endpoint,
        method: options.dynamicSource.method,
        valuePath: options.dynamicSource.valuePath,
        labelPath: options.dynamicSource.labelPath,
        refreshOn: options.dynamicSource.refreshOn
          ? [...options.dynamicSource.refreshOn]
          : [],
      };

      // Copy headers
      if (options.dynamicSource.headers) {
        optionsCopy.dynamicSource.headers = {
          ...options.dynamicSource.headers,
        };
      }

      // Copy parameters
      if (options.dynamicSource.parameters) {
        optionsCopy.dynamicSource.parameters = {
          ...options.dynamicSource.parameters,
        };
      }
    }

    // Handle dependent options
    if (options.dependency) {
      // Process dependent options based on context
      const dependentField = options.dependency.field;
      const dependentValue = this.getValueFromContext(context, dependentField);

      optionsCopy.dependency = {
        field: dependentField,
      };

      // Copy value map
      if (options.dependency.valueMap) {
        optionsCopy.dependency.valueMap = {} as any;

        for (const [key, values] of Object.entries(
          options.dependency.valueMap,
        )) {
          optionsCopy.dependency.valueMap[key] = (values as any[]).map(
            (opt) => ({
              value: opt.value,
              label: opt.label,
              icon: opt.icon,
            }),
          );
        }

        // Filter options based on dependent field value
        if (dependentValue !== null && dependentValue !== undefined) {
          const valueStr = String(dependentValue);
          if (options.dependency.valueMap[valueStr]) {
            optionsCopy.static = options.dependency.valueMap[valueStr].map(
              (opt: any) => ({
                value: opt.value,
                label: opt.label,
                icon: opt.icon,
              }),
            );
          }
        }
      }
    }

    return optionsCopy;
  }

  /**
   * Gets a value from the context using dot notation
   * @param context Context data
   * @param path Dot notation path
   */
  private getValueFromContext(context: Record<string, any>, path: string): any {
    return this.validator.getValueByPath(context, path);
  }
}
